library/modules/EventManager.cpp

```diff
@@ -566,12 +555,40 @@ static void manageJobCompletedEvent(color_ostream& out) {
             seenJobs.emplace(job.id, Job::JobUniquePtr(Job::cloneJobStruct(&job, true)));
         }
         /*
-         * We still need to push back all jobs to maintain the invariant of the algorithm used with
-         * prevJobs and nowJobs. See comments below for more details.
+         * We still need to push back all jobs to maintain the invariant of the
+         * algorithm used with prevJobs and nowJobs.
+         *
+         * Consider a list of job IDs from the job list, including those
+         * that haven't started:
+         * 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
+         *
+         * We push back all jobs, including those that haven't started or have
+         * been assigned yet. Jobs that haven't started or finished have a
+         * completion_timer of -1.
+         *
+         * If we don't push back all jobs to the vector, we could encounter
+         * this situation:
+         *
+         * prevJob IDs (jobs with completion_timer != -1):
+         * 2, 4, 8, 10
+         *
+         * nowJobs IDs (completion_timer != -1 or had completion_timer != -1):
+         * 1, 2, 4, 8, 10
+         *
+         * In this case, Job with ID 1 has started after jobs 2, 4, 8, and 10.
+         * But, nowJobs is not greater than or equal to prevJobs because the ID
+         * 1 in nowJobs is less than the smallest ID in prevJobs,
+         * which breaks the algorithm.
          */
         nowJobs.emplace_back(job.id, job.completion_timer, (bool)job.flags.bits.repeat);
     }
 
+    // Do we want this check?
+    //assert(std::is_sorted(nowJobs.begin(), nowJobs.end()));
```

Comment:
I like the idea of this being in a (Lua) unit test that will run when in fortress mode. The check can be in cpp, but callable from Lua via something like `dfhack.internal.checkInvariants()`